core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala

```diff
@@ -0,0 +1,283 @@
+        && (acl.host == host || acl.host == Acl.WildCardHost)) {
+        authorizerLogger.debug(s"operation = $operations on resource = $resource from host = $host is $permissionType.name based on acl = $acl")
+        return true
+      }
+    }
+    false
+  }
+
+  override def addAcls(acls: Set[Acl], resource: Resource): Unit = {
+    if (acls == null || acls.isEmpty)
+      return
+
+    val updatedAcls: Set[Acl] = getAcls(resource) ++ acls
+    val path: String = toResourcePath(resource)
+
+    if (ZkUtils.pathExists(zkClient, path))
+      ZkUtils.updatePersistentPath(zkClient, path, Json.encode(Acl.toJsonCompatibleMap(updatedAcls)))
+    else
+      ZkUtils.createPersistentPath(zkClient, path, Json.encode(Acl.toJsonCompatibleMap(updatedAcls)))
+
+    inWriteLock(lock) {
+      aclCache.put(resource, updatedAcls)
+    }
+
+    updateAclChangedFlag(resource)
+  }
+
+  override def removeAcls(aclsTobeRemoved: Set[Acl], resource: Resource): Boolean = {
+    if (!ZkUtils.pathExists(zkClient, toResourcePath(resource)))
+      return false
+
+    val existingAcls: Set[Acl] = getAcls(resource)
+    val filteredAcls: Set[Acl] = existingAcls.filter((acl: Acl) => !aclsTobeRemoved.contains(acl))
+    if (existingAcls.equals(filteredAcls))
+      return false
+
+    val path: String = toResourcePath(resource)
+    ZkUtils.updatePersistentPath(zkClient, path, Json.encode(Acl.toJsonCompatibleMap(filteredAcls)))
+
+    inWriteLock(lock) {
+      aclCache.put(resource, filteredAcls)
+    }
+
+    updateAclChangedFlag(resource)
+
+    true
+  }
+
+  override def removeAcls(resource: Resource): Boolean = {
+    if (ZkUtils.pathExists(zkClient, toResourcePath(resource))) {
+      ZkUtils.deletePath(zkClient, toResourcePath(resource))
+      inWriteLock(lock) {
+        aclCache.remove(resource)
+      }
+      updateAclChangedFlag(resource)
+      true
+    } else false
+  }
+
+  override def getAcls(resource: Resource): Set[Acl] = {
+    inReadLock(lock) {
+      aclCache.get(resource).getOrElse(Set.empty[Acl])
+    }
+  }
+
+  private def getAclsFromZk(resource: Resource): Set[Acl] = {
+    val aclJson = ZkUtils.readDataMaybeNull(zkClient, toResourcePath(resource))._1
+    aclJson.map(Acl.fromJson).getOrElse(Set.empty)
+  }
+
+  override def getAcls(principal: KafkaPrincipal): Set[Acl] = {
+    val aclSets = aclCache.values
+    var acls = Set.empty[Acl]
+    for(aclSet  <- aclSets) {
+      acls ++= aclSet.filter(acl => acl.principal.equals(principal))
+    }
+    acls.toSet[Acl]
+  }
+
+  private def loadCache: Set[Acl] = {
+    var acls = Set.empty[Acl]
+    val resourceTypes = ZkUtils.getChildren(zkClient, SimpleAclAuthorizer.AclZkPath)
+    for (rType <- resourceTypes) {
+      val resourceType = ResourceType.fromString(rType)
+      for (resources <- ZkUtils.getChildren(zkClient, resourceType.name)) {
+        for (resourceName <- resources)
+          acls ++= getAcls(Resource(resourceType, resourceName.toString))
```

Comment:
getAcls should be getAclsFromZk, right? Could we add a unit test to test the initial loading?